File: src/ui/lib/escher/types/color.cc
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/ui/lib/escher/types/color.h"

#include <src/lib/fxl/strings/string_printf.h>

namespace escher {

std::ostream& operator<<(std::ostream& os, const ColorRgba& c) {
  return os << fxl::StringPrintf("RGBA:%02X%02X%02X%02X", c.r, c.g, c.b, c.a);
}

std::ostream& operator<<(std::ostream& os, const ColorBgra& c) {
  return os << fxl::StringPrintf("BGRA:%02X%02X%02X%02X", c.b, c.g, c.r, c.a);
}

}  // namespace escher

File: src/graphics/lib/compute/mold/src/tile/map/painter/color.rs
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use super::byte_fraction::ByteFraction;

pub const ZERO: Color = Color {
    red: ByteFraction::zero(),
    green: ByteFraction::zero(),
    blue: ByteFraction::zero(),
    alpha: ByteFraction::zero(),
};

pub const BLACK: Color = Color {
    red: ByteFraction::zero(),
    green: ByteFraction::zero(),
    blue: ByteFraction::zero(),
    alpha: ByteFraction::one(),
};

#[repr(C)]
#[derive(Clone, Copy, Debug, Default)]
pub struct Color {
    pub blue: ByteFraction,
    pub green: ByteFraction,
    pub red: ByteFraction,
    pub alpha: ByteFraction,
}

impl Color {
    pub fn swap_rb(self) -> Self {
        Self { blue: self.red, green: self.green, red: self.blue, alpha: self.alpha }
    }

    pub fn to_rgb565(self) -> u16 {
        let red = u16::from((self.alpha * self.red).value());
        let green = u16::from((self.alpha * self.green).value());
        let blue = u16::from((self.alpha * self.blue).value());

        let red = ((red >> 3) & 0b1_1111) << 11;
        let green = ((green >> 2) & 0b111_1111) << 5;
        let blue = (blue >> 3) & 0b1_1111;

        red | green | blue
    }
}
